3단계 - 로또(자동) PR 요청드립니다.

earth-h opened the pull request:
태경님 안녕하세요!

로또 - TDD :: 3단계 - 로또(자동) PR 피드백 요청드립니다.

이번에 이펙티브 자바 책을 보면서, 정적 팩토리 메소드를 사용해보았습니다.
- private final이어도, List나 Map은 수정될 여지가 있다고 판단하여 해당 필드들은 넣거나 조회할 때에 unmodifiableXXX로 진행하였습니다.
  - 위처럼 할 경우, 성능적 이슈가 발생할 수 있지 않을까 하는 생각이 들긴 하였지만, 출력하기 위해 한 번 조회하는 메소드이거나 테스트에서 사용하는 생성자에 대해 불변리스트화 한 것이라 괜찮지 않을까 생각했습니다(혹 이 부분에 오류가 있다면 피드백 부탁드립니다!!)

- LottoNumber의 경우 1 ~ 45까지 정해진 숫자값만 사용하기 때문에 처음부터 1 ~ 45까지의 숫자를 캐싱해서 사용하고자 구현하였습니다. 
  - 이 부분은 저번 단계에서 피드백주셨던 Patterns처럼 비싼 객체는 아니겠지만, 재사용해도 될 객체라고 판단하여 이렇게 구현하였습니다.

감사합니다 !

earth-h committed:
refactor(utils): StringUtils 내 split() 메소드 개선

StringUtils
- split 메소드 호출 시마다 매번 Pattern 인스턴스를 생성하지 않도록, static final로 선언하여 재사용하도록 함
- matcher.group()에서 사용할 index 값을 각각 static final로 선언하여 어떤 의미를 갖는 값인지 알기 쉽도록 리팩토링

earth-h committed:
docs(guide): README.md 업데이트

step3 로또(자동) 기능 구현 목록 정의

earth-h committed:
feat(domain): LottoNumber 클래스 추가

LottoNumber
- 로또 각 숫자 하나를 의미하는 클래스
- 1 ~ 45 사이의 값을 갖는지 validation
- test코드에서 객체 자체를 비교하기 위해 equals와 hashCode override

ErrorCode
- 로또 과제에서 사용할 ErrorCode를 모으기 위한 enum 클래스 생성

earth-h committed:
refactor(constant): ErrorCode 메시지 수정

ErrorCode
- 로또의 각 자리수 > 로또의 각 숫자로 에러코드 변경

earth-h committed:
feat(domain): Lotto 클래스 추가

Lotto
- 로또는 중복되지 않은 숫자 6개로 이루어짐
- 로또 각각의 숫자가 1~45 사이인지 판단할 수 있도록 Lotto에 선언된 lottoNumbers는 List<LottoNumber>으로 타입을 지정함
- 로또의 각 숫자들에 중복을 확인하고자, List를 Set으로 변환하여 크기를 비교함

ErrorCode
- Lotto 구현으로 인해 추가된 errorCode 추가

earth-h committed:
feat(domain): Money 클래스 추가

Money
- 입력된 금액에 따라 최대로 구매할 수 있는 로또 개수 판단
- 로또 1장 가격인 1000원보다 작은 값이 들어올 경우 validation

Lotto
- 로또 1장 가격을 static final 변수로 관리

ErrorCode
- Money 구현으로 인해 추가된 errorCode 추가

earth-h committed:
feat(domain): LottoPrize 클래스 추가 및 Lotto 역할 추가

LottoPrize
- 당첨로또와 구매로또 간 일치한 개수에 따라 서로 다른 enum값을 반환
- findLottoPrize의 반환값을 Optional이 아닌 LottoPrize로 반환하기 위해 orElse()문 추가
  - 이에 따라, FIRST~FOURTH, NO_PRIZE로 enum 변경

Lotto
- 구매로또와 당첨로또 간 동일한 숫자 개수 판단 메소드 구현
  - 동일한 숫자 있을 경우 1씩 증가하며, 동일하지 않으면 0 증가하도록 구현
  - 이때, 0과 1을 좀 더 명확히 표현하기 위해 static final 변수로 나타냄

earth-h committed:
feat(domain): Lottos 클래스 추가 및 Money 역할 추가

Lottos
- 주어진 금액에 따라 로또들을 구매하여 가지고 있을 클래스 구현

Money
- 수익음을 계산하기 위한 findProfit() 메소드 구현
  - int형인 money를 소숫점이 생기도록 double로 형변환
  - 소숫점 두번째자리까지 나타내도록 소숫점 세째자리 버림

earth-h committed:
refactor(domain): LottoNumber 정적 팩토리 메소드 생성

LottoNumber
- LottoNumber는 각각 1~45 사이의 숫자로, Money와 같이 재사용해도 됨
- Money는 범위가 큰 반면에, LottoNumber는 1~45까지만 존재하므로, lottoNumberCache맵을 생성하여 클래스 생성 시, 각각의 LottoNumber을 Map의 value로 넣어서 사용
- 이에, LottoNumber private 생성자에서는 따로 validation 처리 하지 않고, LottoNumber를 외부에서 꺼내갈 때 가져가려고 하는 lottoNumber가 유효한 값인지 체크하도록 함

earth-h committed:
refactor(constant): ErrorCode 관리를 한 곳에서 하기 위해 common 패키지 내 구성

ErrorCode
- lotto와 calculator 패키지 내에 각각 존재하던 ErrorCode 한 곳으로 모음
  - 중복된 ErrorCode 막기 위함

earth-h committed:
feat(utils): IntegerUtils 클래스 생성

IntegerUtils
- calculator와 lotto 과제 모두에서 사용되는 텍스트의 parseInt 시, 커스텀 에러 발생시키는 메소드 구현

earth-h committed:
test(utils): IntegerUtils 테스트 코드 작성

IntegerUtilsTest
- IntegerUtils의 parseInt 성공/실패 케이스 테스트 코드 작성

earth-h committed:
feat(domain): 로또 숫자 발급을 위한 generator 생성 및 적용

LottoNumberGenerator
- LottoNumber 리스트 생성 시 사용할 generator interface
  - 사용자가 직접 입력하여 lottoNumber 리스트를 만들 수도 있고, shuffle()을 통해 만들수도 있기 때문에 interface 생성

ReadLineLottoNumberGenerator
- 사용자로부터 입력받은 String을 ,(쉼표) 단위로 split하여 lottoNumbers 생성
  - 이때, ,(쉼표) 사이에 빈 공간이 있을 수 있어 trim() 메소드 사용

RandomLottoNumberGenerator
- Collections.shuffle() 메소드를 통해 랜덤한 lottoNumbers 반환
  - shuffle() 메소드는 기존재하는 collection을 뒤섞는 것이기 때문에 static final로 생성된 lottoNumbers에 LottoNumber가 될 수 있는 객체를 모두 넣어둠
- shuffle() 후 로또에 필요한 개수만큼 추출한 후 sort()

LottoNumber
- RandomLottoNumberGenerator에 의해 sort()할 수 있도록 Comparable<LottoNumber>를 implements하고, compareTo()를 오버라이딩

Lotto
- LottoNumberGenerator 생성을 통해 로또 생성 방법 변경

earth-h committed:
refactor(domain): 메소드명 및 파라미터명 리팩토링

Lotto
- prizeLotto의 경우 LottoPrize와 헷갈릴 여지가 있어 winningLotto로 변경

Money
- 수익금과 수익금비율 각각의 메소드가 있어야 해 비율을 나타내는 메소드명 findProfit > findProfitsRatio 변경

earth-h committed:
feat(domain): 로또 결과 반환을 위한 LottoResults 클래스 생성

LottoResults
- 당첨금 총 합 및 각 LottoPrize별 개수 확인을 위해 LottoResults 클래스 내 Map<LottoPrize, Integer> 생성

Lottos
- 테스트코드에서 사용할 수 있도록, List<Lotto> lottos를 인자로 가지는 생성자 추가
- 로또 결과를 가지는 LottoResults를 반환하는 로또 결과 생성 메소드 추가

earth-h committed:
refactor(domain): toString 메소드 리팩토링

LottoNumber
- Lotto에 속한 LottoNumber들을 리스트형식으로 출력할 수 있도록 toString 메소드 변경

Lotto
- Lotto를 출력할 때, Lotto가 가진 lottoNumbers를 출력하도록 toString 메소드 오버라이딩

earth-h committed:
refactor(domain): Money 클래스의 수익률 계산 메소드 리팩토링

Money
- findProfitsRatio 메소드의 주체와 비교 대상이 헷갈릴 여지가 있어 docs 추가 및 네이밍 수정

earth-h committed:
feat(domain): Money 클래스 내 비교 메소드 추가

Money
- 수익금과 로또 구매시 사용한 돈 간의 비교를 위해 isLessThan 메소드 추가

earth-h committed:
refactor(domain): Lottos 클래스 내 unmodifiableList 적용

Lottos
- Lottos(List<Lotto> lottos) 생성자는 테스트 시에만 사용하지만, public으로 열려있어 누군가 해당 생성자를 이용해 lottos를 생성할 수 있으므로, this.lottos에 list 넣을 때 불변 리스트로 넣도록 변경
  - 이펙티브 자바 아이템 15. 클래스의 멤버의 접근 권한을 최소화하라에서 길이가 0이 아닌 배열은 모두 변경 가능함을 보고 리팩토링
- 로또 리스트 출력을 위해 List<Lotto>를 외부에 반환해주어야 하는데, 이때 Lottos 클래스 내 lottos가 변경되지 않도록 unmodifiableList로 반환함

earth-h committed:
feat(domain): LottoResults 내 LottoPrize별 개수 반환 메소드 생성

LottoResults
- findLottoResultCount() 메소드 추가

earth-h committed:
feat(domain): LottoPrize 내 메소드 추가

LottoPrize
- isNoPrize() 메소드 추가
  - 화면 출력 시, NO_PRIZE에 해당하는 항목은 출력하지 않기 위해 boolean 메소드 추가

earth-h committed:
feat(view): 콘솔 input/output용 view 생성

InputView
- input을 받기 위한 scanner 보관
- input 받기 전 가이드를 print하기 위한 메소드 생성

ResultView
- input에 따라 생성된 결과를 나타낼 메소드 생성

earth-h committed:
feat(controller): LottoController 생성

LottoController
- Lotto 전반적인 내용을 진행하는 컨트롤러
- 사용자 input이 잘못되어 IllegalArgumentException 발생 시, 다시 입력받을 수 있도록 try ~ catch문 사용

earth-h committed:
feat(application): LottoApplication 구현

LottoApplication
- Lotto 게임 수행을 위한 main 메소드를 가진 클래스 생성

earth-h committed:
refactor(domain): LottoResults 생성자 리팩토링

LottoResults
- LottoResults의 Map<LottoPrize, Integer>에 바로 접근 가능한 public 생성자를 private로 변경한 후 정적 팩토리 메소드를 통해 접근하도록 변경
  - 이펙티브 자바 아이템 15. 클래스의 멤버의 접근 권한을 최소화하라에서 길이가 0이 아닌 배열은 모두 변경 가능함을 보고 리팩토링
  (Map 역시 참조변수이므로, put이 추가로 일어날 여지가 있다고 보고, unmodifiableMap으로 들어온 인자를 감싸서 LottoResults 생성)

Lottos
- LottoResults 생성자 대신 정적 팩토리 메소드 사용하여 생성하도록 리팩토링

earth-h committed:
refactor(domain): Money 메소드 위치 리팩토링

Money
- overriding 메소드를 클래스 가장 하단에 두고자 메소드 위치 리팩토링

earth-h committed:
docs(guide): README.md 업데이트

step3 로또(자동) 구현 후, 완료된 내용 업데이트

earth-h committed:
fix(domain): Money 정적 팩토리 메소드 반영

Money
- 기존에 Money 생성시 로또 1개 금액보다 작으면 validation하는 부분이 있었음
  - 이 경우, 수익금 계산 시 수익이 없으면 0이 Money 생성자에 들어가 exception 발생됨 확인
  - 이에, Money 생성 시 사용할 메소드를 2가지로 나누어 사용하고자 함
- Lotto 구매 시 사용하는 createLottoMoney()
- 일반 돈 계산 시 사용하는 createMoney()

ohtaeg commented on README.md:
```diff
@@ -71,5 +71,84 @@
+8. [x] "가,2:3"
+
+## 3단계 - 로또(자동)
+### 기능 목록
```
꼼꼼하게 잘 작성해주셨네요 👍 
작성된 내용을 잘 반영해주셨구요 👍

ohtaeg commented on src/main/java/calculator/utils/StringUtils.java:
```diff
@@ -7,12 +7,15 @@ public class StringUtils {
 
     private static final String BASIC_DELIMITER_REGEX = ",|:";
     private static final String CUSTOM_DELIMITER_REGEX = "//(.)\n(.*)";
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile(CUSTOM_DELIMITER_REGEX);
+    private static final int CUSTOM_DELIMITER_PATTERN_SPLIT_DELIMITER_INDEX = 1;
+    private static final int CUSTOM_DELIMITER_PATTERN_SPLIT_TARGET_INDEX = 2;
```
피드백 반영 👍

ohtaeg commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,68 @@
+    public static final int MIN_LOTTO_NUMBER = 1;
+    public static final int MAX_LOTTO_NUMBER = 45;
+
+    private static final Map<Integer, LottoNumber> lottoNumberCache = new HashMap<>();
```
> LottoNumber의 경우 1 ~ 45까지 정해진 숫자값만 사용하기 때문에 처음부터 1 ~ 45까지의 숫자를 캐싱해서 사용하고자 구현하였습니다.

너무 잘해주셨네요 💯

ohtaeg commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Objects;
+import common.constant.ErrorCode;
+
+public class LottoNumber implements Comparable<LottoNumber> {
```
로또 번호를 작게 잘 분리해주셨어요 👍 
잘 구현해주셔서 크게 드릴 피드백이 없네요 👍

ohtaeg commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private void validateDupliteLottoNumbers(List<LottoNumber> lottoNumbers) {
+        Set<LottoNumber> numbers = new HashSet<>(lottoNumbers);
```
Set을 이용한 중복 제거 👍

ohtaeg commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -0,0 +1,64 @@
+
+public class ResultView {
+
+    private static final String PURCHASING_LOTTO_COUNT = "%d개를 구매했습니다.\n";
```
사소하지만 OS 마다 줄바꿈을 인식하는게 다르기 때문에 jdk 1.7부터 도입된 현재 OS에 맞는 줄바꿈을
얻어오도록 하면 좋을 것 같아요! `System.lineSeperator();`

ohtaeg commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public LottoPrize findLottoPrize(Lotto winningLotto) {
```
이 부분은 같이 고민해보면 좋을 것 같아요 :)

Lottos 객체도 LottoPrize를 의존하고 있고, Lotto 객체도 LottoPrize를 의존하고 있는 상황인데요!
개인적으로 당첨을 판단하는 로직은 Lottos에서만 해주도록 한다면 아래와 같은 장점이 있을 것 같다고 생각이 듭니다..!

당첨에 대한 의존을 Lottos만 관심을 가질 수 있게 되면 6개의 로또 번호(Lotto)가 가진 책임을 명확하게 할 수 있을 것 같으며
Lotto와 당첨에 대한 의존성을 최소화 할 수 있는 것 같아요!

Lotto와 LottoPrize 의존성을 제거해주면 좋을 것 같은데 이에 대해 지수님은 어떻게 생각하시는지 궁금하네요 😄

earth-h replied:
코드를 구현할 때에는, 막연하게 Lottos는 Lotto의 집합을 가지고 있으니, Lotto 각각이 몇등인지는 Lotto 클래스 가서 해야한다고 생각을 했는데요!

태경님이 말씀해주신것처럼 Lotto와 LottoPrize 의존성을 제거하는 것이 당첨등수 찾기 로직 요구사항이 변경되어도 수월하게 변경할 수 있을 것 같습니다 :) 

피드백 주신 부분 반영해보도록 하겠습니다. 감사합니다 !

earth-h replied:
태경님! 혹시,, 힌트를 얻어볼 수 있을까요~?

말씀해주신것처럼 의존을 끊어보고자 하는데, findLottoPrize() 메소드는 아무리 생각해도 Lotto에 있어야 할 거 같아서요😭
단일 로또별로 prize를 결정하고 있어서, 위 구조를 어떻게 바꿀 수 있을지 방안이 잘 떠오르지 않아 여쭤봅니다..!

ohtaeg replied:
제가 생각했던 의견을 말씀드려보면, 
Lotto에서 당첨 갯수만 Lottos에게 알려주고 Lottos가 당첨 갯수로 prize를 결정해도 괜찮지 않을까 했습니다!
만약 해당 피드백 반영이 힘드실 경우 현재 구조를 유지해주셔도 좋습니다 :)

earth-h replied:
답변 감사합니다! 

생각을 엉뚱하게 하고 있었네요..! Lotto와 LottoPrize간의 관계만 끊으면 되니, count 넘긴 후애 lottos에서 LottoPrize를 만들어주면 되는 것을 캐치하지 못했습니다 😭

ohtaeg replied:
아닙니다! 구현을 잘해주셔서 크게 드릴 피드백들이 없네요 👍 
늦은 밤까지 열심히 하시는 모습이 귀감이 되네요 👍 
화이팅입니다 🔥

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public Lottos(List<Lotto> lottos) {
+        this.lottos = Collections.unmodifiableList(lottos);
```
이 부분은 같이 얘기해보면 좋을 것 같아요 :)

생성자에서 불변하도록 관리해도 좋지만, 보통 객체 내에서의 상태 값들은 해당 객체에서만 변경되고 사용하기 위함인데요 :)
즉, 내부에서만 상태 값들을 자주 사용할 것이기에 자체적으로 불변하게 관리를 안해도 되지 않을까? 라는 의견이에요!
외부에 제공해야할 경우 구현 해주신 것 처럼 read-only로 리턴하여 외부에서 값의 변조를 막아도 되지 않을까합니다!

물론 정답이 없는 부분이기에 지수님의 의견이 궁금하여 코멘트를 드려봅니다 😄

earth-h replied:
말씀해주신 생성자는, 테스트코드 작성 시 편하게 하기 위해 생성한 생성자입니다 :) 

저는 어플리케이션 로직상에서는 해당 생성자를 사용하지는 않았지만, 
만약 누군가 저 생성자를 이용하여서 Lottos를 생성한다고 가정하였을 때 파라미터로 받은 lottos를 외부에서 변경하면
변경된 내용이 Lottos에 반영이 되지 않을까하여서 작성해두었는데요!

혹 위와 같은 상황일때 `this.lottos = lottos;`로 작성하여도 이슈가 없을까요~?

ohtaeg replied:
아하! 외부에서 lottos를 넘겨주었을 때 외부에서 계속 레퍼런스 참조를 할 경우를 대비하신거군요 👍 
엄청 꼼꼼하게 잘 방어해주셨네요 :)

Collections.unmodifiableList(lottos);를 사용하더라도 unmodifiableList() 내부 구현을 보면
그대로 레퍼런스를 사용하고 있기에 언제든지 lottos의 변조가 가능할 것 같아요 :)
```java
public static <T> List<T> unmodifiableList(List<? extends T> list) {
        return (list instanceof RandomAccess ?
                new UnmodifiableRandomAccessList<>(list) :
                new UnmodifiableList<>(list));
}

// UnmodifiableList class
UnmodifiableList(List<? extends E> list) {
    super(list);
    this.list = list;
}
```
이런 경우, `복사 생성자`를 이용하는 방법은 어떨지 조심스럽게 의견을 전달드려봅니다 😄 
인터넷에 레퍼런스가 많아서 복사 생성자를 검색하신다음 마음에 드시는 레퍼런스를 참조하시면 좋을 것 같아 링크는
따로 드리지 않는 부분 양해 부탁드립니다 🙏  
(드릴려는 레퍼런스가 따로 있었는데 검색이 안되어서 못드리겠네요 😭 )

earth-h replied:
조회용으로 리턴할 때에는 `Collections.unmodifiableList(___);`로 사용하면, 조회하는 영역에서 해당 리스트를 변경할 수 없겠지만, 생성자에서 저렇게 넣으면 외부에서는 여전히 변경이 가능하겠네요! 

답변 감사합니다 ㅎㅎ 
말씀해주신 부분 참고하여서, 아래와 같이 수정진행하도록 하겠습니다 :) 

```java
    public Lottos(List<Lotto> lottos) {
        this.lottos = new ArrayList<>(lottos);
    }
```

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,46 @@
+        List<Lotto> lottos = new ArrayList<>();
+        int maxLottoCount = money.maxLottoCount();
+        for(int i = 0; i < maxLottoCount; i++) {
+            lottos.add(Lotto.generateLotto(new RandomLottoNumberGenerator()));
```
사소하지만 Lotto 외부에서 생성 방법(RandomLottoNumberGenerator)을 주입해주는 것 처럼
Lottos 외부에서도 생성 방법을 주입시켜주면 어떨까요 :) ?
외부에서 Lottos 생성 요청을 할 때 어떻게 생성할지도 정해서 주면 괜찮을 것 같아서요!
게다가 Lottos 테스트할 때도 용이할 것 같구요 :)

earth-h replied:
Controller에서부터 생성 방법을 주입해오는 것이, 좀 더 관리할 때 편할 것 같네요!
해당 주입받은 내용을 그대로 lotto에 반영하면 될거 같아서 Controller에서 Lottos로 넣고, Lottos에서 Lotto로 넣는 방식으로 진행하겠습니다 !

Lottos 같은 경우Lotto 리스트를 생성하여 테스트를 해야해서, 아래와 같은 생성자로 테스트 시 사용하고 있습니다!
테스트코드를 입출력으로 받기에도, 랜덤값으로 받기에도 애매한터라, 생성자를 따로 추가해두었습니다 :) 
```java
    public Lottos(List<Lotto> lottos) {
        this.lottos = new ArrayList<>(lottos);
    }
```

ohtaeg replied:
**수정을 위한 리뷰가 아니오니** 참고만 해주시면 감사하겠습니다 :)

테스트를 위한 생성자 관련해서 매우 주관적이고 개인적인 의견을 공유 드리고자 해서요!
저는 테스트에서 사용되는 생성자를 선호하는 편은 아니긴 합니다 :)
해당 생성자가 없으면 객체가 안전하다는 등의 보장을 못하거나 테스트 하기 힘들 경우에는 만드는 편입니다.

선호하지 않는 이유는 코드는 테스트를 위해 존재하는게 아닌 실제 운영환경에 맞게 객체가 생성되는게 바람직하다고 생각합니다 :)
코드가 있어서 테스트를 할 수 있는거라  운영환경에서 필요한 상황에 맞게 생성자를 통해 메모리에 적재가 되어야한다고 생각하는데요!

그렇기에  위와 같은 경우 꼭 생성자가 아니더라도 setter 등의 방법을 통해 언제든지 로또 생성 전략을 바꿀 수 있도록 해줘도 괜찮지 않을까 합니다 😄 

물론 생성자 추가도 좋습니다 👍 해당 피드백은 개인적인 의견을 공유드리는게 목적이라 참고만 해주세요 ㅎㅎ

ohtaeg commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,86 @@
+    private static final double FLOOR_LOCATION = 100;
+    private static final int MIN_MONEY = 0;
+
+    private final int money;
```
사소하지만 당첨 금액이 21억이 넘을 수도 있지 않을까 해요 😄

ohtaeg commented on src/main/java/lotto/domain/LottoResults.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Collections;
+import java.util.Map;
+
+public class LottoResults {
```
👍 👍

ohtaeg commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.Arrays;
+
+public enum LottoPrize {
```
당첨 정보를 enum으로 관리 💯

ohtaeg commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -0,0 +1,45 @@
+
+public enum LottoPrize {
+
+    NO_PRIZE(0, 0, "0~2개 일치 (0원)"),
```
이 부분은 같이 고민해보면 좋을 것 같아요 :)

lottoPrizeMessage는 비지니스 데이터라기 보다는 출력을 위한 데이터에 더 가깝다고 생각이 되어지는데요!
도메인 패키지를 만들어주신 의도에 맞게 핵심 비지니스 로직을 가지는 객체에서
어떻게 그려지는지에 대한 관심은 분리되면 좋지 않을까 합니다 :)

해당 작업은 View 에서 이루어지면 좋을 것 같은데 지수님은 어떻게 생각하시나요?

ohtaeg commented on src/main/java/lotto/domain/RandomLottoNumberGenerator.java:
```diff
@@ -0,0 +1,34 @@
+
+public class RandomLottoNumberGenerator implements LottoNumberGenerator {
+
+    private static final List<LottoNumber> lottoNumbers = new ArrayList<>();
+
+    static {
+        for(int lottoNumber = MIN_LOTTO_NUMBER; lottoNumber <= MAX_LOTTO_NUMBER; lottoNumber++) {
+            lottoNumbers.add(LottoNumber.from(lottoNumber));
+        }
+    }
```
이미 LottoNumber에서 캐시를 관리하고 있는데 제너레이터에서 따로 캐싱해주시는 의도가 있으신가요 :) ?

earth-h replied:
shuffle을 하기 위해서 로또 번호 전체를 List로 넣고자 하였습니다!

해당 리스트를 만들 때, 캐싱되어 있는 LottoNumber를 사용하기 위해 lottoNumbers라는 List에는 캐싱된 lottoNumber를 넣고 있습니다.

LottoNumber에서는 Map에 캐싱을 하여서 각 key(로또번호)에 맞는 LottoNumber를 가져오는 방식으로 되어 있어서, shuffle을 하기 위해 추가적으로 List를 생성하였던 부분입니다. 

혹, 위 부분 대신에 LottoNumber 클래스에 하기 메소드 작성 후에 RandomLottoNumberGenerator 클래스의 generateLottoNumbers()에서 호출해서 쓰는 것이 나은 방식일까요? 
아래 방식으로 하게 되면, 매번 generator 호출시마다 list 생성이 이루어질것으로 보여 다시 한 번 캐싱한 것이긴 합니다..! 

```java
    public static List<LottoNumber> getCachingLottoNumbers() {
        return new ArrayList<>(lottoNumberCache.values());
    }
```

ohtaeg replied:
아하 이 부분은 제가 오해했었네요 🙇  두군데에서 캐싱하는 줄 알았습니다 🙏 
친절한 설명 감사드립니다!

말씀해주신대로 LottoNumber에서 캐싱된 객체들을 전달받기 때문에 크게 상관없을 것 같아요 👍

ohtaeg commented on src/test/java/lotto/domain/LottoPrizeTest.java:
```diff
@@ -0,0 +1,13 @@
+
+import org.junit.jupiter.api.Test;
+
+public class LottoPrizeTest {
```
해당 도메인에 대한 테스트가 조금 부족한 것 같아요 🙏

ohtaeg reviewed (changes requested):
늦어서 죄송합니다 🙇 
3단계 구현도 잘해주셨네요 👍 👍 전체적으로 구조를 잘 잡아주셨어요!
다음 미션을 진행하기전에 같이 고민해보면 좋을 포인트들이 존재하여 Request changes 하였습니다 🙏
피드백 확인 후 반영하여 재요청 주시면 감사하겠습니다 :)
좋은 주말, 좋은 밤 되시길 바랄께요!

earth-h committed:
refactor(domain,view): domain에 존재하던 출력을 위한 데이터 view로 리팩토링

LottoPrize
- 핵심 비즈니스 로직만을 가지도록, 출력과 관련된 lottoPrizeMessage 제거

ResultView
- LOTTO_RESULT 생성 시, LottoPrize를 통해 가져온 값들로 출력물 생성 리팩토링

earth-h committed:
refactor(view): 줄바꿈 처리 변경

ResultView
- Window, Linux간 서로 다른 개행 방식을 사용함에 따라 \n 대신 System.lineSeparator() 사용하도록 로직 수정

earth-h committed:
test(domain): LottoPrize 테스트 코드 추가 작성

LottoPrizeTest
- isNoPrize())에 대한 테스트코드 추가
- findLottoPrize()에 대한 테스트코드 추가

earth-h committed:
refactor(domain): Money의 int money > long money 형 변환

피드백주신대로, 복권 구매 시 사용자가 입력하는 금액이나 복권 총 당첨금액이 int형을 넘어서는 경우를 대비해 long 형으로 변경

Money
- int money > long money

LongUtils
- Money의 money를 long으로 변경함에 따른 형변환 메소드 추가

earth-h committed:
refactor(domain): Lotto와 LottoPrize간 의존 끊기 리팩토링

기존에 Lottos - LottoPrize, Lotto - LottoPrize 각각 의존성이 있어 Lottos - LottoPrize간 의존만 남도록 로직 변경

Lotto
- Lotto에서 LottoPrize를 반환하던 findLottoPrize의 반환값을 int로 변경하여 LottoPrize 생성 시 필요한 matchCount를 반환하도록 변경
  - 반환타입이 변경됨에 따라 메소드 명도 리팩토링(findLottoMatchCount)

Lottos
- Lotto로부터 matchCount를 받아오므로, 해당 count값을 이용한 LottoPrize 생성

earth-h committed:
fix(domain): Lottos 생성자 리팩토링

Lottos
- Lottos 생성자 내 Collections.unmodifiableList()를 사용한 이유는 외부에서 생성자를 통해 Lottos를 생성할 때에, 변경을 막기 위함이었음
  - 하지만, 외부에서는 여전히 해당 List에 참조하고 있어 변경이 가능함
  - 이에, 복사생성자로 변경

earth-h committed:
refactor(domain, controller): Lottos 생성자 파라미터에 LottoNumberGenerator 추가

LottoController
- Lottos 생성 시, LottoNumberGenerator 주입을 Controller에서 하여 관리 포인트 줄이고자 함

Lottos
- Lottos 생성자 파라미터에 LottoNumberGenerator를 넣어, List<Lotto> 생성 시, Lotto에 해당 generator 보냄

ohtaeg commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -4,21 +4,19 @@
 
 public enum LottoPrize {
 
-    NO_PRIZE(0, 0, "0~2개 일치 (0원)"),
```
피드백 반영 👍

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -14,27 +14,28 @@ public class Lottos {
     private final List<Lotto> lottos;
 
-    public Lottos(Money money) {
+    public Lottos(Money money, LottoNumberGenerator lottoNumberGenerator) {
```
🙇 👍

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -14,27 +14,28 @@ public class Lottos {
         Map<LottoPrize, Integer> lottoResults = new HashMap<>();
         for(Lotto lotto: lottos) {
-            LottoPrize lottoPrize = lotto.findLottoPrize(winningLotto);
+            int matchCount = lotto.findLottoMatchCount(winningLotto);
+            LottoPrize lottoPrize = LottoPrize.findLottoPrize(matchCount);
```
피드백 반영 👍

ohtaeg commented on src/test/java/common/LongUtilsTest.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class LongUtilsTest {
```
새로 생긴 부분에 대해서 꼼꼼하게 테스트 👍

ohtaeg commented on src/test/java/lotto/domain/LottosTest.java:
```diff
@@ -8,7 +8,7 @@ public class LottosTest {
 
     @Test
```
createLottoResults()에 대한 테스트도 추가해주시면 좋을 것 같아요 :)

ohtaeg reviewed (approved):
피드백 반영을 잘해주셨어요 👍 
생성자 관련해서 추가 코멘트를 드렸는데 확인해주시면 감사하겠습니다 :)
새벽까지 고생많으셨습니다 🌃 
다음 미션도 화이팅입니다 🔥
